### PR TITLE
Add SslConfig APIs and fix SNI and SSL hostname bugs

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
The server SSL config APIs for SNI didn't allow configuring SSL
associated with each individual SNI host. The client APIs coupled the
host name verification algorithm with the non-authoritative peer
host/port. This doesn't correctly represent the semantics of SSL.

Modifications:
- Deprecate all APIs related to SecurityConfigurator and secure()
  builder method. The secure() builder methods require that users call
  commit() which is easy to miss if the builder is used in a conditional
  block, and prototypes of a composable server SNI API results in code
  that is difficult to read and understand which configurations apply.
- Add SslConfig interface types and builders that can be used to create
  the ssl configuration independent of the protocol builders. This
  allows for folks to configure ssl independently and we can overlay any
  values which maybe automatically inferred in the protocol builder
  (e.g. peerHost on the client, alpn protocols, etc.).
- Server SSL allows for providing SNI configuration, and Netty's
  DomainWildcardMappingBuilder is used to match hostnames.
- Client side SSL config object removes methods which couple the host
  name verification algorithm to the non-authoritative peer host/port,
  and adds new methods which allow setting them independently.
- Update all tests to enable host name verification, and avoid
  deprecated method use when possible.

Result:
SSL config APIs now correctly represent SNI semantics for client and
server. The client SSL config decouples the non-authoritative peer
host/port from the verification algorithm.

The deprecated types and methods will be removed in a future release.

### CODE REVIEW

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/PartitionedHttpClientSecurityConfigurator.java
Comment (bondolo) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/PartitionedHttpClientSecurityConfigurator.java:27]: Missing advice

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DelegatingSslConfig.java
Comment (bondolo) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DelegatingSslConfig.java:49]: Is it expected that these would return a fresh InputStream for each invocation or the same stream? Trying to decide if config is one-shot or reusable and it is unclear.

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/MultiAddressHttpClientBuilder.java
Comment (Scottmitch) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/MultiAddressHttpClientBuilder.java:113]: instead of adding a specific callback for ssl configuration, I created this more general callback to configure the builder (long discussed task). may still be worth cleaning up specific overrides later now that this general mechanism exists.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpServerConfig.java
Comment (bondolo) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpServerConfig.java:64]: Presumably uses LinkedHashMap for iteration order but that is not the type of sniMapOverrides or sniMap. Why is it a LinkedHashMap?

**Comment by bondolo** [comment]
I like the new API.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpServerConfig.java
Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpServerConfig.java:64]: Yes LHM is to preserve iteration order. I'll add a comment.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DelegatingSslConfig.java
Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DelegatingSslConfig.java:49]: I assume you are asking about the base `SslConfig` interface? Since ownership is transferred (e.g. caller must invoke `close()`) it must be a new instance on each invocation of the `Supplier`. Currently these methods will be invoked at most 1 time per ssl setup. Do you feel the ownership transfer or other impacts should be made more explicit in the javadocs?

Comment (bondolo) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DelegatingSslConfig.java:49]: I would expect SslConfig to be reused for many requests as it seems expensive (reading keys and certs) to construct. The test example shows it being created inline which is dangerous in that we don't provide other examples and someone may copy it and use it in that way in their code where they should be creating one and reuse it many times. (The peril of using tests as sample code).

Otherwise, the other part is whether a "currying" Supplier is permitted, that can return the same stream instance every time or whether a true supplier that returns a "fresh" stream each time is required.  That a Supplier is used rather than providing the stream directly, or something else like contents of the stream might imply that the supplier is invoked more than once. The docs about the stream being closed by the impl are good and helpful but since close() is idempotent and it is not an error to call close many times it doesn't really answer whether the Supplier needs to be reusable.

Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DelegatingSslConfig.java:49]: > reused for many requests ... doesn't really answer whether the Supplier needs to be reusable.

The config object will be re-used for each ssl setup/handshake (e.g. connection initialization) and I'll clarify in the comments.

> we don't provide other examples

There are a few examples independent from unit tests which demonstrate `SslConfig` usage (`Http*MutualTLS`, `Http*WithAlpn`), and inline creation is expected to be safe.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpServerConfig.java
Comment (bondolo) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpServerConfig.java:65]: Perhaps sniMap should be a LinkedHashMap as well for the same reason?

**Review by bondolo — APPROVED** [review]

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpServerConfig.java
Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpServerConfig.java:65]: I don't think we need to restrict which map the user is able to provide. If iteration ordering is important they can always provide LHM, but maybe there is no overlap/wildcards and they don't need the additional overhead.

**Review by idelpivnitskiy — CHANGES_REQUESTED** [review]
Great that the new approach solves our issues, like it!

File: servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/DefaultClientSslConfigBuilder.java
Comment (Scottmitch) [reply servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/DefaultClientSslConfigBuilder.java:54]: some form of trust manager is "required" it just so happens the default constructor falls back to the JVM's default trust manager, which may not be appropriate for all use cases (e.g. default trust manager emulates web browser behavior if not overridden). Due to its relative priority I think it is worth calling it out explicitly rather than having it blend in with the other setters (especially when contrasted against the server side that requires the key manager in the constructor).

File: servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/DefaultClientSslConfigBuilder.java
Comment (Scottmitch) [reply servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/DefaultClientSslConfigBuilder.java:133]: Will do. Noe that SSLEngine doesn't validate this value and the accessor which returns the value states that [the port number of the peer, or -1 if nothing is available](https://github.com/openjdk/jdk8u/blob/master/jdk/src/share/classes/javax/net/ssl/SSLEngine.java#L431), and looks if `-1` [the session cache isn't populated](https://github.com/openjdk/jdk8u/blob/master/jdk/src/share/classes/sun/security/ssl/SSLSessionContextImpl.java#L183).

File: servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/DefaultClientSslConfigBuilder.java
Comment (Scottmitch) [reply servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/DefaultClientSslConfigBuilder.java:144]: Will do.

File: servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/SslConfig.java
Comment (Scottmitch) [reply servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/SslConfig.java:110]: sure I'll just use `List`.

File: servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/DefaultServerSslConfigBuilder.java
Comment (Scottmitch) [reply servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/DefaultServerSslConfigBuilder.java:106]: I'm not to keen on force copying if its not necessary and if the user wants/needs to re-use the collection they can always pass a copy.

File: servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/AbstractSslConfigBuilder.java
Comment (Scottmitch) [reply servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/AbstractSslConfigBuilder.java:224]: done

File: servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/SslConfig.java
Comment (Scottmitch) [reply servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/SslConfig.java:102]: for HTTP there are a number of params we try to automatically infer, and ALPN is one of them. however the ALPN parameters are a function of the ssl configuration and don't have to be inferred by all protocols and this now unifies the SSL config API at the TCP level (e.g. `TcpClientConfig`, `TcpServerConfig`) to be focused on `SslConfig` instead of independent params.

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/MultiAddressHttpClientBuilder.java
Comment (Scottmitch) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/MultiAddressHttpClientBuilder.java:113]: > There is no hint for users how to customize SslConfig for multi-address client.

The method provides a `@deprecation` link for the replacement method. I also added a `@Deprecated` annotation on `unresolvedAddressToHost(..)` method which should only be applicable at the `SingleAddressHttpClientBuilder` level. Can you clarify if there are other methods that need to be covered? I still think we should move toward the "configure the single address your self based upon scheme/address" instead of a special case for ssl configuration.

```java
    /**
     * Sets a function that is used for configuring SSL/TLS for https requests.
     * @deprecated Use {@link #appendSingleAddressConfigurator(SingleAddressConfigurator)}.
     * @param sslConfigFunction The function to use for configuring SSL/TLS for https requests.
     * @return {@code this}
     */
    @Deprecated
    public abstract MultiAddressHttpClientBuilder<U, R> secure(
            BiConsumer<HostAndPort, ClientSecurityConfigurator> sslConfigFunction);
```

Comment (idelpivnitskiy) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/MultiAddressHttpClientBuilder.java:113]: Current description for deprecated `secure` method is great and covers everything. I tried to say that when we remove this deprecated method, there are no more hints for users how to customize the ssl config.

Long term, if we remove all other methods (like, `enableWireLogging`, `socketOption`, etc.) and keep only `singleAddressConfigurator`, that will help users to understand all single-address related configuration is in one place. But then we won't be able to extend `HttpClientBuilder` anymore. While we have a mix of `HttpClientBuilder` methods with `singleAddressConfigurator`, it's hard to discover how to customize the ssl config.

File: servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/SslConfig.java
Comment (idelpivnitskiy) [reply servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/SslConfig.java:102]: I understand it simplifies some internal API, but complicates public api and understanding of ssl config. I bet users will try to configure this list for http and will do that incorrectly (inconsistently with http protocols). Because it's additive, let's defer it until we have a valid use-case when this configuration is required. For now, it's easier and safer for users if we infer alpn protocols.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/AlpnServerContext.java
Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/AlpnServerContext.java:56]: I will restore this assert. The current code allows overriding the ALPN protocols via SslConfig. The idea was folks could re-order their ALPN preference for each individual SNI map entry on the server, but todo this safely we would need to coordinate the http configuration with the SNI selection as well. since we don't need this now I'll punt.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:264]: summary of offline discussion:
- Deprecate all methods in multi/partitioned builders that are specific to individual client config
- Users can use the “client configurator” method to customize options for each client.
- Adjust/simplify client builder hierarchy as required (e.g. we may not be able to extend the same classes).
- Remove Deprecated methods in followup release

File: servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/AbstractSslConfigBuilder.java
Comment (Scottmitch) [reply servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/AbstractSslConfigBuilder.java:64]: good idea. done.

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/MultiAddressHttpClientBuilder.java
Comment (Scottmitch) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/MultiAddressHttpClientBuilder.java:129]: discussed offline and summarized result/approach here: https://github.com/apple/servicetalk/pull/1387#discussion_r586075555

**Comment by Scottmitch** [comment]
build failure attributed to https://github.com/apple/servicetalk/issues/1224

File: servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/SslConfig.java
Comment (Scottmitch) [reply servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/SslConfig.java:103]: I've updated the http code to always take the inferred value and clarified in the builder that the set value maybe ignored (depending upon the protocol).

**Comment by Scottmitch** [comment]
build failure attributed to https://github.com/apple/servicetalk/issues/999

File: servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/DefaultServerSslConfigBuilder.java
Comment (Scottmitch) [reply servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/DefaultServerSslConfigBuilder.java:31]: I'll remove the `Default` from the name (shorter name, and alphabetically next to the type it builds) and add a `@see` to the javadoc.

File: servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/AbstractSslConfigBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/AbstractSslConfigBuilder.java:276]: Does this param need any validation?

File: servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/AbstractSslConfigBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/AbstractSslConfigBuilder.java:292]: Does this param need any validation?

File: servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/AbstractSslConfigBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/AbstractSslConfigBuilder.java:307]: `requireNonNull`?

**Review by idelpivnitskiy — APPROVED** [review]
LGTM after the last comments addressed:

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/MultiAddressHttpClientBuilder.java
Comment (Scottmitch) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/MultiAddressHttpClientBuilder.java:129]: I agree. I would like to update the `PartitionHttpClientBuilderConfigurator` API to use a consistent pattern that is used here, and update the method name too.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:265]: The default behavior hasn't changed. Folks that make no code changes should see no behavior changes (e.g.  `secure()` works the same way it does today). If folks start using the new API, they will need to transition over to it (e.g. no half-in / half-out).

The `configurator` API is similar to `SchemeRegistry` AHC uses to register socket factories. We don't have an explicit `Scheme` type but the usage would be somewhat awkward here because `HostAndPort` is the second argument.

```java
SchemeRegistry registry = new SchemeRegistry();
registry.register(new Scheme("http", PlainSocketFactory.getSocketFactory(), 80));
registry.register(new Scheme("https", SSLSocketFactory.getSocketFactory(), 443));
```

The JDK client allows for sending requests to different URLs with the same client, but AFAICT it doesn't allow configuring different trust managers per request. The "one single default setting" works well for "browser" use cases but doesn't work cross trust domains in service to service use cases. In these cases you would have to create an independent client, which is what multi-address is doing under the hood (we just allow them to configure all the options).

```java
HttpClient c = HttpClient.newBuilder().sslContext(/*SSLContext*/).build();
HttpRequest req = HttpRequest.newBuilder(URI.create("https://www.servicetalk.io")).GET().build();
HttpResponse<T> response = c.send(req, /*response body handler*/);
```

Lets sync up offline to resolve this.

Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:265]: Most users don't touch the `secure()` method at all, because they are satisfied with the default ssl config we provide for `https` scheme. With new API, their default behavior should remain the same, they should not care about invoking `sslConfig` now. This use-case should continue to work:
```java
BlockingHttpClient client = HttpClients.forMultiAddressUrl().buildBlocking();
client.request(client.get("https://servicetalk.io"));
```
Users who care about custom configuration, can use configurator to customize it.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java
Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java:596]: > Should we worry about : being the last character

I don't think this is necessary. if the hostname is malformed we will throw.

> square brakets around IPv6 address?

sure I can update the code to handle `[<host>]:<port>`

File: servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/AbstractSslConfigBuilder.java
Comment (Scottmitch) [reply servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/AbstractSslConfigBuilder.java:307]: sure I can add it here. `null` is the default state so folks could conceivably "set back to default" ... but we can allow this later if we need to.

File: servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/ClientSslConfigBuilder.java
Comment (Scottmitch) [reply servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/ClientSslConfigBuilder.java:139]: I'll defer on enforcing bounds here bcz the usage currently doesn't enforce limits (other than `-1` if "not available"). I'll defer to the underlying layers to apply limits is necessary.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:265]: ah ok, I see what you are saying. thought the primary concern was related to passing the `scheme` in the configurator. I will refactor to preserve the "default trust manager always applied for https schema"

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java
Comment (Scottmitch) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:217]: https://github.com/apple/servicetalk/issues/1423
https://github.com/apple/servicetalk/issues/1424

**Pull request merged**
